adcc/AdcMatrix.py

```diff
@@ -307,28 +313,153 @@ def __init__(self, method, hf_or_mp):
+        if block_orders["ph_pphh"] != block_orders["pphh_ph"]:
+            raise ValueError("ph_pphh and pphh_ph should always have "
+                             "the same order")
+        if block_orders["ph_pphh"] is not None \
+           and block_orders["pphh_pphh"] is None:
+            raise ValueError("pphh_pphh cannot be node if ph_pphh isn't.")
+
+        if intermediates is None:
+            self.intermediates = Intermediates(self.ground_state)
+        else:
+            self.intermediates = intermediates
+
+        # Build the blocks
+        variant = None
+        if method.is_core_valence_separated:
+            variant = "cvs"
+        self.__blocks = {
+            block: ppmatrix.block(self.ground_state, block.split("_"), order=order,
+                                  intermediates=self.intermediates,
+                                  variant=variant)
+            for block, order in block_orders.items() if order is not None
+        }
+
+        # TODO Remove this
+        # automatically get the shape and subspace
+        # blocks of the matrix from the orders
+        # TODO once the AmplitudeVector object support missing blocks
+        #      this would be self.diagonal.blocks and self.diagonal.ph.subspaces
+        #      and self.diagonal.pphh.subspaces
+        self.__block_spaces = {}
+        if variant == "cvs":
+            self.__block_spaces["s"] = ["o2", "v1"]
+            if block_orders.get("pphh_pphh", None) is not None:
+                self.__block_spaces["d"] = ["o1", "o2", "v1", "v1"]
+        else:
+            self.__block_spaces["s"] = ["o1", "v1"]
+            if block_orders.get("pphh_pphh", None) is not None:
+                self.__block_spaces["d"] = ["o1", "o1", "v1", "v1"]
+        self.blocks = list(self.__block_spaces.keys())
 
-    def compute_matvec(self, in_ampl, out_ampl=None):
+    @timed_member_call()
+    def diagonal(self, block):
+        if block not in self.blocks:
+            raise ValueError("block not in blocks")
+        # TODO The block argument should be brought more
+        #      in line with the ph, pphh stuff
+
+        # TODO Once the AmplitudeVector object support missing blocks this would
+        #      just be return sum(bl.diagonal for bl in self.__blocks)
+
+        if block == "s":
+            return evaluate(sum(bl.diagonal.ph
+                                for bl in self.__blocks.values()
+                                if bl.diagonal and "s" in bl.diagonal.blocks
+                                and bl.diagonal.ph))
+        else:
+            return evaluate(sum(bl.diagonal.pphh
+                                for bl in self.__blocks.values()
+                                if bl.diagonal and "d" in bl.diagonal.blocks
+                                and bl.diagonal.pphh))
+
+    def compute_apply(self, block, tensor):
+        # TODO A lot of junk code to get compatibility to the old interface
```

Comment:
Since the old interface is gone, can we remove the `s` and `d` labels and replace them by `ph` and `pphh`, respectively?

Reply:
That is the point of the two remaining checkboxes in the PR overview ;)